roadmap/implementers-guide/src/node/availability/availability-recovery.md

```diff
@@ -26,30 +26,144 @@ Output:
+    validator_index: ValidatorIndex,
+    candidate_hash: CandidateHash,
+    response: ResponseChannel<ChunkResponse>,
+}
+
 struct State {
-    /// The interaction handle may just be the interaction itself, but it might be useful to implement each interaction
-    /// as its own `Future`, running in its own task, in which case the handle will be a channel to communicate with
-    /// the interaction.
+    /// Each interaction is implemented as its own async task, and these handles are for communicating with them.
     interactions: Map<CandidateHash, InteractionHandle>,
-    live_requests: Map<RequestId, (PeerId, CandidateHash)>,
+    /// A recent block hash for which state should be available.
+    live_block_hash: Hash,
+    discovering_validators: Map<AuthorityDiscoveryId, Vec<AwaitedChunk>>,
+    live_chunk_requests: Map<RequestId, (PeerId, AwaitedChunk)>,
     next_request_id: RequestId,
+    connecting_validators: Stream<(AuthorityDiscoveryId, PeerId)>,
+
+    /// interaction communication. This is cloned and given to interactions that are spun up.
+    from_interaction_tx: Sender<FromInteraction>,
+    /// receiver for messages from interactions.
+    from_interaction_rx: Receiver<FromInteraction>,
 
     // An LRU cache of recently recovered data.
-    availability_lru: LruCache<CandidateHash, AvailableData>,
+    availability_lru: LruCache<CandidateHash, Result<AvailableData, RecoveryError>>,
+}
+
+struct InteractionHandle {
+    awaiting: Vec<ResponseChannel<Result<AvailableData, RecoveryError>>>,
+}
+
+struct Unavailable;
+enum FromInteraction {
+    // An interaction concluded.
+    Concluded(CandidateHash, Result<AvailableData, RecoveryError>),
+    // Make a request of a particular chunk from a particular validator.
+    MakeRequest(
+        AuthorityDiscoveryId, 
+        CandidateHash, 
+        ValidatorIndex, 
+        ResponseChannel<ChunkResponse>,
+    ),
+    // Report a peer.
+    ReportPeer(
+        PeerId,
+        Rep,
+    ),
 }
-```
 
-```rust
 struct Interaction {
-    responses: ResponseChannel<Option<AvailableData>>,
+    to_state: Sender<FromInteraction>,
     validator_authority_keys: Vec<AuthorityId>,
     validators: Vec<ValidatorId>,
     // a random shuffling of the validators which indicates the order in which we connect to the validators and
     // request the chunk from them.
     shuffling: Vec<ValidatorIndex>, 
     // The number of pieces needed.
     threshold: usize, 
+    candidate_hash: Hash,
+    erasure_root: Hash,
     received_chunks: Map<ValidatorIndex, ErasureChunk>,
-    requesting_chunks: Set<ValidatorIndex>,
+    requesting_chunks: FuturesUnordered<Receiver<ChunkResponse>>,
 }
-```
\ No newline at end of file
+```
+
+### Signal Handling
+
+On `ActiveLeavesUpdate`, if `activated` is non-emtpy, set `state.live_block_hash` to the first block in `Activated. Also, remove any items that have been in 
+
+Ignore `BlockFinalized` signals.
+
+On `Conclude`, shut down the subsystem.
```

Comment:
Is the node shutting down here? We'd prefer if recovery finished for whatever assignments we annouced, but doing that can be future work.

Reply:
Yeah, this is the node shutting down. I think we'll file this as 'future work': https://github.com/paritytech/polkadot/issues/2029

Reply:
I think we'd best handle this on the level of other subsystems - it's really the requesting subsystem that knows best if a particular availability recovery interaction can be safely cancelled or not. I think this subsystem should be ready to conclude at any time, whereas the approval voting subsystem will have to be aware of the recovery requests it has in-flight.